main/src/library/Channel/Channel.flix

```diff
@@ -0,0 +1,582 @@
+    */
+    pub opaque type Channel[a] = (
+        Int64,
+        ReentrantLock,
+        Int,
+        Bool,
+        MutList[a],
+        MutList[LockConditionPair],
+        Condition
+    )
+
+    /**
+    * public int branchNumber;
+    * public boolean defaultChoice;
+    * public Object element;
+    */
+    pub opaque type SelectChoice[a] = (
+        Int,
+        Bool,
+        a
+    )
+
+    pub def newSelectChoice(branchNumber: Int, element: a, defaultChoice: Bool): SelectChoice[a] =
+        SelectChoice(
+            branchNumber,
+            defaultChoice,
+            element
+        )
+    // FYI default choice is false in this constructor, TODO : delete
+
+    // channel, index
+    pub opaque type ChannelIndexPair[a] = (
+        Channel[a],
+        Int
+        )
+
+    pub def newChannelIndexPair(channel: Channel[a], index: Int): ChannelIndexPair[a] =
+        ChannelIndexPair(
+            channel,
+            index
+        )
+
+    pub def getChannel(cip: ChannelIndexPair[a]): Channel[a] = match cip {
+            case ChannelIndexPair(c, _)   => c
+        }
+
+    pub def getIndex(cip: ChannelIndexPair[a]): Int = match cip {
+            case ChannelIndexPair(_, ind)   => ind
+        }
+
+    /// @Override
+    /// pub def equals(o: a): Bool & Impure =
+    ///     if (this == o) return true;
+    ///     if (o == null || getClass() != o.getClass()) return false;
+    ///     ChannelIndexPair that = (ChannelIndexPair) o;
+    ///     return index == that.index && Objects.equals(channel, that.channel);
+    /// }
+
+    /// @Override
+    /// pub def hashCode(cip: ChannelIndexPair): Int =
+    ///     import java.util.Objects.hash as hash;
+    ///     hash(getChannel(cip), getIndex(cip))
+
+
+    pub def getId(c: Channel[a]): Int64 = match c {
+            case Channel(id, _, _, _, _, _, _)   => id
+        }
+
+    pub def getChannelLock(c: Channel[a]): ReentrantLock = match c {
+            case Channel(_, channelLock, _, _, _, _, _)   => channelLock
+        }
+
+    pub def getBufferSize(c: Channel[a]): Int = match c {
+            case Channel(_, _, bufferSize, _, _, _, _)   => bufferSize
+        }
+
+    pub def getUnbuffered(c: Channel[a]): Bool = match c {
+            case Channel(_, _, _, unbuffered, _, _, _)   => unbuffered
+        }
+
+    pub def getElementQueue(c: Channel[a]): MutList[a] = match c {
+            case Channel(_, _, _, _, elementQueue, _, _)   => elementQueue
+        }
+
+    // TODO : make it a MutList or MutSet??
+    pub def getWaitingGetters(c: Channel[a]): MutList[LockConditionPair] = match c {
+            case Channel(_, _, _, _, _, waitingGetters, _)   => waitingGetters
+        }
+
+    pub def getWaitingSetters(c: Channel[a]): Condition = match c {
+            case Channel(_, _, _, _, _, _, waitingSetters)   => waitingSetters
+        }
+
+
+    /*
+    instance Order[Channel] {
+        pub def compare(x: Channel, y: Channel): Comparison =
```

Comment:
Channel should be `ChannelImpl[a]` everywhere (note the missing type variable in your code).